File: app/models/rag_eval.py
"""RAG evaluation lab models."""

from __future__ import annotations

import uuid
from typing import Any

from sqlalchemy import Boolean, Float, ForeignKey, Index, Integer, String, Text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, TimestampMixin


class RagGoldenQuery(Base, TimestampMixin):
    """Admin-owned golden query for retrieval/evidence evaluation."""

    __tablename__ = "rag_golden_queries"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    query: Mapped[str] = mapped_column(Text, nullable=False)
    expected_answer: Mapped[str | None] = mapped_column(Text, nullable=True)
    expected_source_ids: Mapped[list[str]] = mapped_column(JSONB, nullable=False, default=list)
    tags: Mapped[list[str]] = mapped_column(JSONB, nullable=False, default=list)
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    created_by_user_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)

    __table_args__ = (
        Index("ix_rag_golden_queries_created_by_user_id", "created_by_user_id"),
        Index("ix_rag_golden_queries_is_active", "is_active"),
        Index("ix_rag_golden_queries_created_at", "created_at"),
    )


class RagEvalRun(Base, TimestampMixin):
    """One execution of a golden query against the current RAG evidence engine."""

    __tablename__ = "rag_eval_runs"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    golden_query_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("rag_golden_queries.id", ondelete="SET NULL"), nullable=True
    )
    query: Mapped[str] = mapped_column(Text, nullable=False)
    status: Mapped[str] = mapped_column(String(32), nullable=False, default="pending")
    score: Mapped[float] = mapped_column(Float, nullable=False, default=0.0)
    evidence_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    missing_expected_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    result_json: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=False, default=dict)
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_by_user_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)

    __table_args__ = (
        Index("ix_rag_eval_runs_golden_query_id", "golden_query_id"),
        Index("ix_rag_eval_runs_status", "status"),
        Index("ix_rag_eval_runs_created_by_user_id", "created_by_user_id"),
        Index("ix_rag_eval_runs_created_at", "created_at"),
    )

File: front/shared/types/rag-eval.ts
export interface RagGoldenQueryCreate {
  name: string
  query: string
  expected_answer?: string | null
  expected_source_ids: string[]
  tags: string[]
}

export interface RagGoldenQueryResponse {
  id: string
  name: string
  query: string
  expected_answer?: string | null
  expected_source_ids: string[]
  tags: string[]
  is_active: boolean
  created_by_user_id: string
  created_at: string
  updated_at: string
}

export interface RagGoldenQueryListResponse {
  items: RagGoldenQueryResponse[]
  total: number
  page: number
  page_size: number
}

export interface RagEvalRunCreate {
  golden_query_id?: string | null
  query?: string | null
}

export interface RagEvalRunResponse {
  id: string
  golden_query_id?: string | null
  query: string
  status: string
  score: number
  evidence_count: number
  missing_expected_count: number
  result_json: Record<string, unknown>
  error_message?: string | null
  created_by_user_id: string
  created_at: string
  updated_at: string
}

export interface RagEvalRunListResponse {
  items: RagEvalRunResponse[]
  total: number
  page: number
  page_size: number
}
